packages/swing-store/src/transcriptStore.js

```diff
@@ -235,6 +304,106 @@ export function makeTranscriptStore(
+    SELECT endPos
+    FROM transcriptSpans
+    WHERE vatID = ? AND startPos = ?
+  `);
+  sqlGetSpanEndPos.pluck(true);
+
+  const sqlReadSpanItems = db.prepare(`
+    SELECT item
+    FROM transcriptItems
+    WHERE vatID = ? AND ? <= position AND position < ?
+    ORDER BY position
+  `);
+
+  /**
+   * Read the items in a transcript span
+   *
+   * @param {string} vatID  The vat whose transcript is being read
+   * @param {number} [startPos] A start position identifying the span to be
+   *    read; defaults to the current span, whatever it is
+   *
+   * @returns {IterableIterator<string>}  An iterator over the items in the indicated span
+   */
+  function readSpan(vatID, startPos) {
+    /** @type {number | undefined} */
+    let endPos;
+    if (startPos === undefined) {
+      ({ startPos, endPos } = getCurrentSpanBounds(vatID));
+    } else {
+      insistTranscriptPosition(startPos);
+      endPos = sqlGetSpanEndPos.get(vatID, startPos);
+      if (typeof endPos !== 'number') {
+        throw Fail`no transcript span for ${q(vatID)} at ${q(startPos)}`;
+      }
+    }
+    startPos <= endPos || Fail`${q(startPos)} <= ${q(endPos)}}`;
+    const expectedCount = endPos - startPos;
+
+    function* reader() {
+      let count = 0;
+      const compressedSpanReader = readCompressedSpan(vatID, startPos);
+      if (compressedSpanReader) {
+        for (const item of compressedSpanReader()) {
+          yield item;
+          count += 1;
+        }
+      } else {
+        for (const { item } of sqlReadSpanItems.iterate(
+          vatID,
+          startPos,
+          endPos,
+        )) {
+          yield item;
+          count += 1;
+        }
+      }
+      count === expectedCount ||
+        Fail`read ${q(count)} transcript entries (expected ${q(
+          expectedCount,
+        )})`;
+    }
+
+    if (startPos === endPos) {
+      return empty();
+    }
+
+    return reader();
+  }
+
+  const sqlAddCompressedSpan = db.prepare(`
+    INSERT INTO transcriptCompressedSpans (vatID, startPos, endPos, items, incarnation)
+    VALUES (?, ?, ?, ?, ?)
+  `);
+
+  const sqlDeleteSpanItems = db.prepare(`
+    DELETE FROM transcriptItems
+    WHERE vatID = ? AND position >= ? AND position < ?
+  `);
+
+  function compressSpan(vatID, startPos, endPos, incarnation) {
+    // XXX It's not clear to me what the most efficient way to produce the
+    // compressed item blob is, but the following at least works. It should
+    // suffice unless and until we decide we need something faster or with
+    // a lower memory footprint, but since individual spans are modestly sized, I
+    // don't immediately see that as a pressing issue.  Note that the obvious move
+    // of switching to an implementation based on Streams would cause this
+    // function to become async, which would in turn bubble up into public APIs
+    // that are already in use which expect synchronous behavior, and hence such
+    // a change would be quite invasive.
```

Comment:
Synchronous is nearly always safer and simpler if you can get away with it.  Since the SQLite calls themselves are synchronous, we only introduce asynchrony when we need to stream a relatively larger volume of data accessed across multiple SQL queries.   I actually started out implementing this piece as async, on the theory that we already had perfectly good data streaming code that did this in the snapstore, but it quickly turned into a big tarbaby so I backed off to sync unless and until circumstances demonstrate that async is required.